### DISCUSSION TITLE
How should pointer dwell work?

### DISCUSSION THREAD

**Discussion by gbishop** [UNC-Project-Open-AAC/OS-DPI#117] [Q&A]
Our current implementation of pointer dwell does not work properly. I'm trying to think about how it should work in some edge cases.

Consider a grid with 2 buttons named A and B.

In the Method tab we have access to two raw signals from the browser: Pointer Enter and Pointer Exit. We allow the designer to specify a debounce interval on each of them. The signals we will pass along to the Method after debouncing I'll call simply Enter and Exit. The Method Responses are trigger by those.

Now please consider the table below. The first column is the time step. These are some arbitrary interval. The Pointer Enter debounce is set to 3 time steps and the Pointer Exit debounce is set to 6 time steps. I'm not claiming these are good values; I'm using them for this thought experiment. 

The next two columns show the raw events. The A in column 2 of line 1 means we got a Pointer Enter event from the A button. The A
in column 3 of line 2 means we got a Pointer Exit event from the A button. 

On line 7 we signal the Enter event (for processing by our responses) corresponding to the raw Pointer Enter event that occurred on line 4.

Lines 8 through 21 show that we could leave button A for less than 6 time steps and not signal Exit.

On line 22 we finally signal Exit on button A.

On line 23 we enter A again and on 26 we trigger its Enter actions. 

We then leave A and enter B. But we are not going to execute the Enter actions for B until we execute the Exit actions for A in step 33.

Finally! I come to my question on line 34. **Should we immediately trigger the button B Enter actions?** We've been on button B for 6 time steps but we held off triggering Enter on B because we hadn't finished the debounce interval for A.

|  step  | Pointer Enter | Pointer Exit | Enter | Exit |   Comments                                  |
| -- | --------- | -------- | ----- | ---- | ----------------------------------- |
| 1  | A         |          |       |      |                                     |
| 2  |           | A        |       |      | ignored too brief                   |
| 3  |           |          |       |      |                                     |
| 4  | A         |          |       |      |                                     |
| 5  |           |          |       |      |                                     |
| 6  |           |          |       |      |                                     |
| 7  |           |          | A     |      | long enough                         |
| 8  |           | A        |       |      | ignored too brief                   |
| 9  | A         |          |       |      |                                     |
| 10 |           |          |       |      |                                     |
| 11 |           | A        |       |      | ignored too brief                   |
| 12 | B         |          |       |      |                                     |
| 13 |           |          |       |      |                                     |
| 14 |           | B        |       |      |                                     |
| 15 | A         |          |       |      |                                     |
| 16 |           | A        |       |      |                                     |
| 17 |           |          |       |      |                                     |
| 18 |           |          |       |      |                                     |
| 19 |           |          |       |      |                                     |
| 20 |           |          |       |      |                                     |
| 21 |           |          |       |      |                                     |
| 22 |           |          |       | A    | finally left A for 6 ticks          |
| 23 | A         |          |       |      |                                     |
| 24 |           |          |       |      |                                     |
| 25 |           |          |       |      |                                     |
| 26 |           |          | A     |      | long enough                         |
| 27 |           | A        |       |      | leave A                             |
| 28 | B         |          |       |      | enter B                             |
| 29 |           |          |       |      |                                     |
| 30 |           |          |       |      |                                     |
| 31 |           |          |       |      |        |
| 32 |           |          |       |      |                                     |
| 33 |           |          |       | A    | signal that we left A after 6 ticks |
| 34 |           |          | B?     |      | do we immediately signal enter B?   |

**Comment by lageist** [comment]
Thanks @gbishop. I think we would want to immediately trigger the button B Enter actions if that debounce interval is met. But I'm also struggling to think of a use case that would have the Exit debounce drive the selection method by ignoring the next Enter that meets the specified dwell time. Can you help with that?

**Comment by gbishop** [comment]
No, I can't imagine a situation where the exit debounce should hold off the
enter event but if it doesn't then what does exit debounce mean? I'm not
sure anyone would (should?) set the exit debounce higher than the enter
debounce but if they did, what should happen?

We could say that Enter events take precedence and will cancel any Exit
debounce that might be happening. I guess that could apply with two buttons
having a gap between. In this case you could exit A without entering B.
Then the Exit debounce should apply. But if you then entered B and remained
long enough, the B Enter events would trigger even if the A Exit debounce
had not expired.

Perhaps that would be more natural?

At the level of the Methods tab, I'm trying to allow you to do anything.
Maybe you aren't even using the Enter event and are triggering on Exit
only? I'd like to make it possible for designers to do unanticipated
things.

**Comment by lageist** [comment]
I think Enter taking precedence is right, at least for the use cases I can think of at the moment, but appreciate the goal of supporting designers to do unanticipated things here. Thinking edge cases, thought experiment like, and maybe someone using a head mouse wants to explore Exit debounce being set to longer than Enter to create a delay (long enough to look up and gain eye contact with a partner) before triggering an event like Speak? So composition might be something like Enter + Enter (canceling Exit debounce) + Enter (canceling Exit debounce) + Enter (and then look up at communication partner moving head mouse off display) to trigger Speak when Exit debounce time has elapsed. Would that be implemented by adjusting Enter/Exit? I guess taking this case further, maybe with a page of utterance-based messages someone wants to dwell on each one for a longer period of time to read what's there, and like you said not even use the Enter event, and the Exit time is what triggers the specified action, again looking up at a partner and moving the pointer off screen is the scenario that comes to mind here. We do not need to be thinking about the "Touch" Exit/Release case here, right? Fairly sure that's handled as a different method than what you're asking about. Where a user positions a physical pointer (e.g., hand, finger, head stylus) on the display and has improved access/stability by dragging on the display and having selections accepted only on Exit/release.
